component/otelcol/connector/spanlogs/consumer.go

```diff
@@ -0,0 +1,282 @@
+
+		if !logSpans && !logRoots && !logProcesses {
+			return nil
+		}
+
+		if logProcesses {
+			keyValuesProcesses := pcommon.NewMap()
+
+			c.processKeyVals(keyValuesProcesses, rs, serviceName)
+
+			// Add a trace ID to the key values
+			keyValuesProcesses.PutStr(c.opts.overrides.TraceIDKey, traceID)
+
+			lastTraceID = traceID
+			err := c.appendLogRecord(typeProcess, keyValuesProcesses, logRecords)
+			if err != nil {
+				return err
+			}
+		}
+
+		// Construct the key values
+		keyValues := pcommon.NewMap()
+
+		c.spanKeyVals(keyValues, span)
+		c.processKeyVals(keyValues, rs, serviceName)
+
+		// Add a trace ID to the key values
+		keyValues.PutStr(c.opts.overrides.TraceIDKey, traceID)
+
+		if logSpans {
+			err := c.appendLogRecord(typeSpan, keyValues, logRecords)
+			if err != nil {
+				return err
+			}
+		}
+
+		if logRoots {
+			err := c.appendLogRecord(typeRoot, keyValues, logRecords)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func copyOtelMap(out pcommon.Map, in pcommon.Map, pred func(key string, val pcommon.Value) bool) {
+	in.Range(func(k string, v pcommon.Value) bool {
+		if pred(k, v) {
+			newVal := out.PutEmpty(k)
+			v.CopyTo(newVal)
+		}
+		return true
+	})
+}
+
+func convertOtelMapToSlice(m pcommon.Map) []any {
+	res := make([]any, 0, m.Len())
+
+	m.Range(func(k string, v pcommon.Value) bool {
+		res = append(res, k)
+		res = append(res, v.AsRaw())
+		return true
+	})
+
+	return res
+}
+
+func (c *consumer) appendLogRecord(kind string, keyValues pcommon.Map, logRecords plog.LogRecordSlice) error {
+	newLogRecord, err := c.createLogRecord(kind, keyValues)
+	if err != nil {
+		return err
+	}
+
+	logRecord := logRecords.AppendEmpty()
+	newLogRecord.MoveTo(logRecord)
+	return nil
+}
+
+func (c *consumer) createLogRecord(kind string, keyValues pcommon.Map) (*plog.LogRecord, error) {
+	// Create an empty log record
+	res := plog.NewLogRecord()
+
+	// Add the log line
+	keyValuesSlice := convertOtelMapToSlice(keyValues)
+	logLine, err := logfmt.MarshalKeyvals(keyValuesSlice...)
+	if err != nil {
+		return nil, fmt.Errorf("unable to marshal keyvals due to error: %w", err)
+	}
+
+	if logLine != nil {
+		res.Body().SetStr(string(logLine))
+	}
+
+	// Add the attributes
+	logAttributes := res.Attributes()
+
+	// Add logs instance label
+	logAttributes.PutStr(c.opts.overrides.LogsTag, kind)
+
+	copyOtelMap(logAttributes, keyValues, func(key string, val pcommon.Value) bool {
+		// Check if we have to include this label
+		_, ok := c.opts.labels[key]
+		return ok
+	})
+
+	return &res, nil
+}
+
+func (c *consumer) processKeyVals(output pcommon.Map, resource pcommon.Resource, svc string) {
+	rsAtts := resource.Attributes()
+
+	// Add an attribute with the service name
+	output.PutStr(c.opts.overrides.ServiceKey, svc)
+
+	for _, name := range c.opts.processAttributes {
+		att, ok := rsAtts.Get(name)
+		if ok {
+			// name/key val pairs
+			val := output.PutEmpty(name)
```

Comment:
Not sure what we are doing here. So if we find a value in `rsAtts`, we put an empty value into the output and then copy the empty value to att?

Reply:
Yeah. It does look weird. The issue is that [pcommon.Map](https://github.com/open-telemetry/opentelemetry-collector/blob/main/pdata/pcommon/map.go) doesn't have a function to insert a `Value` and a `Key` at the same time, like a `PutValue(k string, v Value)`. These kinds of functions only insert specific types like `PutInt(k string, v int64)`.

I want to insert a Value so that I preserve the original type, and this is the only way to do it as far as I can tell.

In hindsight, I should have just used a `map[string]pcommon.Value` instead of a `pcommon.Map`. It would also have been more efficient, since a `pcommon.Map` is actually just an slice. Inserting to it is slow, because it has to traverse the whole slice to see if the element is already in the map.

I'll add a TODO in the code for this, but for now I hope it's fast enough that a `pcommon.Map` will do.